R/check_faers_structure.R

```diff
@@ -0,0 +1,130 @@
+#' Check the structure of the FAERS folder
+#'
+#' Check if exists a folder containing the FAERS data and if the structure of
+#' the folder is correct.
+#'
+#' @param path (chr) The path of the folder containing a subfolder (named
+#' "faers_raw_data") with FAERS data inside, sorted by year and quarter.
+#' @param faersyears (chr) A vector of characters with the list of the years of
+#' FAERS data (defaults: all years available).
+#'
+#' @return (lgl) TRUE (invisible) if the "faers_raw_data" exists and it is
+#' structured correctly, FALSE (invisible) otherwise.
+#' @export
+#'
+#' @examples
+#' check_faers_structure(".")
+check_faers_structure <- function(path, faersyears = years_from_faers_html()) {
+  if (!check_root_folder(path)) return(invisible(FALSE))
+  if (length(check_years_directory(path, faersyears)) > 0L) {
+    print(faers_folder_structure(path))
+    return(invisible(FALSE))
+  }
+  if (check_quarter_every_year(path, faersyears)) {
+    print(faers_folder_structure(path))
+    return(invisible(FALSE))
+  }
+  if (length(check_files(path)) > 0L) {
+    print(faers_folder_structure(path))
+    return(invisible(FALSE))
+  }
+  invisible(TRUE)
+}
+
+
+check_root_folder <- function(path) {
+  dirlist <- list.dirs(path, recursive = FALSE)
+  is_faers_folder <- grepl("faers_raw_data", dirlist)
+  if (sum(is_faers_folder) == 0L) {
+    warning("I can't find a folder named 'faers_raw_data'.")
+    return(FALSE)
+  }
+  if (sum(is_faers_folder) > 1L) {
+    warning("Too Many folders named 'faers_raw_data' (or similar).")
+    return(FALSE)
+  }
+  TRUE
+}
+
+
+check_years_directory <- function(path, faers_years) {
+  subfolders <- list.dirs(
+    paste0(path, "/faers_raw_data"),
+    recursive = FALSE, full.names = FALSE
+  )
+  subfolders_logical <- purrr::map(
+    subfolders,
+    function(x) stringr::str_detect(faers_years, glue::glue("^{x}$"))
+  )
+  bad_folders <- subfolders[which(purrr::map(subfolders_logical, sum) == 0L)]
+  if (length(bad_folders) > 0L) {
+    warning(glue::glue("The following folders do not match a FAERS year: ",
+                       ".../faers_raw_data/{toString(bad_folders)}. ",
+                       "Please remove the folders or change directory path."))
+  }
+  bad_folders
```

Comment:
Attenzione alle notazioni: se con `check_*` stai a indicare qualcosa che restituisce `TRUE` o `FALSE` sii consistente. Qui forse la logica potresti spezzarla in due, uno che controlla, e uno che se serve, valuta quali siano. Ma decidi tu, comunque va bene, in caso mettilo in un issue come cose da fare aggiuntive un domani

Iin generale ti suggerisco di renderlo esplicito con nomi tipo `is_*`, o `are_*`; quando leggo `check_*` mi aspetto che non succeda nulla, o che venga restituito un errore (e va bene se invisibilmente restituisce `TRUE` comunque.

Reply:
Ok, ho fatto in modo che sia consistente e per ogni funzione `check_*` restituisca `TRUE `o `FALSE`. Poi se avanza tempo in un secondo momento si può fare in modo che quelle funzioni restituiscano gli elementi non coerenti con la struttura delle cartelle, ma siccome è superfluo non l'ho fatto ora.